File: src/scripts/archetypes.lua
---@source ~/zig/dev/src/api.zig

---@param ecs userdata
---@param allocator userdata
---@return integer
function SpawnPlayer(ecs, allocator)
   local a = allocator
   local id = api.lvl.newEntity(ecs, a)
   api.lvl.addComponent(ecs, a, id, "physics", [[{
   "pos": {"x": 5, "y": 5}
   }]])
   api.lvl.addComponent(ecs, a, id, "is_player", nil)
   api.lvl.addComponent(ecs, a, id, "sprite", [[
      {"animation_player": {"animation_name": "player"}}
   ]])
   api.lvl.addComponent(ecs, a, id, "inventory", nil)
   api.lvl.addComponent(ecs, a, id, "hitbox", nil)
   api.lvl.addComponent(ecs, a, id, "wall_collisions", nil)
   return id
end

---@param ecs userdata
---@param allocator userdata
---@return integer
function SpawnSlime(ecs, allocator)
   local a = allocator
   local id = api.lvl.newEntity(ecs, a)
   api.lvl.addComponent(ecs, a, id, "physics", [[{
   "pos": {"x": 3, "y": 3}
   }]])
   api.lvl.addComponent(ecs, a, id, "wanderer", nil)
   api.lvl.addComponent(ecs, a, id, "health", nil)
   api.lvl.addComponent(ecs, a, id, "hitbox", nil)
   api.lvl.addComponent(ecs, a, id, "death_particles", nil)
   api.lvl.addComponent(ecs, a, id, "sprite", [[
      {"animation_player": {"animation_name": "slime"}}
   ]])
   api.lvl.addComponent(ecs, a, id, "metadata", [[{
      "archetype": "slime"
   }]])
   api.lvl.addComponent(ecs, a, id, "loot", [[{
      "items": ["SpawnCoin"]
   }]])
   api.lvl.addComponent(ecs, a, id, "wall_collisions", nil)
   return id
end

---@param ecs userdata
---@param allocator userdata
---@return integer
function SpawnMovementParticle(ecs, allocator)
   local a = allocator
   local id = api.lvl.newEntity(ecs, a)
   api.lvl.addComponent(ecs, a, id, "health", nil)
   api.lvl.addComponent(ecs, a, id, "sprite", [[{
      "animation_player": {
         "animation_name": "particle",
         "tint": {"r": 100, "g": 100, "b": 100, "a": 50}
      },
      "z_level": "foreground"
   }]])
   api.lvl.addComponent(ecs, a, id, "health_trickle", nil)
   api.lvl.addComponent(ecs, a, id, "metadata", [[{
      "archetype": "particle"
   }]])
   return id
end

---@param ecs userdata
---@param allocator userdata
---@return integer
function SpawnFireball(ecs, allocator)
      local a = allocator
      local id = api.lvl.newEntity(ecs, a)
      api.lvl.addComponent(ecs, a, id, "sprite", [[{
         "animation_player": {
            "animation_name": "fireball",
            "tint": {"r": 200, "g": 100, "b": 100, "a": 150}
         }
      }]])
      api.lvl.addComponent(ecs, a, id, "hitbox", [[{
         "top": 0.1, "bottom": 0.1, "left": 0.1, "right": 0.1
      }]])
      api.lvl.addComponent(ecs, a, id, "damage", [[{
         "type": "force", "amount": 10
      }]])
      api.lvl.addComponent(ecs, a, id, "health", nil)
      api.lvl.addComponent(ecs, a, id, "health_trickle", nil)
      api.lvl.addComponent(ecs, a, id, "invulnerable", nil)
      api.lvl.addComponent(ecs, a, id, "metadata", [[{
         "archetype": "fireball"
      }]])
      api.lvl.addComponent(ecs, a, id, "death_animation", [[
         {"animation_name": "fireball_explosion"}
      ]])
      return id
end

---@param ecs userdata
---@param allocator userdata
---@return integer
function SpawnCoin(ecs, allocator)

   local a = allocator
   local id = api.lvl.newEntity(ecs, a)
   api.lvl.addComponent(ecs, a, id, "sprite", [[{
      "animation_player": {
         "animation_name": "coin",
         "tint": {"r": 143, "g": 201, "b": 218, "a": 250}
      },
      "z_level": "background"
   }]])
   api.lvl.addComponent(ecs, a, id, "hitbox", [[{
      "top": 0.1, "bottom": 0.1, "left": 0.1, "right": 0.1
   }]])
   api.lvl.addComponent(ecs, a, id, "physics", nil)
   api.lvl.addComponent(ecs, a, id, "item", [[{
   "type": "coin"
   }]])
   api.lvl.addComponent(ecs, a, id, "invulnerable", nil)
   api.lvl.addComponent(ecs, a, id, "metadata", [[{
      "archetype": "item"
   }]])
   return id;
end


---@param ecs userdata
---@param allocator userdata
---@return integer
function SpawnBloodParticle(ecs, allocator)
   local a = allocator
   local id = api.lvl.newEntity(ecs, a)
   api.lvl.addComponent(ecs, a, id, "health", nil)
   api.lvl.addComponent(ecs, a, id, "sprite", [[{
      "animation_player": {
         "animation_name": "particle",
         "tint": {"r": 143, "g": 0, "b": 0, "a": 50}
      },
      "z_level": "background"
   }]])
   api.lvl.addComponent(ecs, a, id, "health_trickle", nil)
   api.lvl.addComponent(ecs, a, id, "metadata", [[{
      "archetype": "particle"
   }]])
   return id
end

---@param ecs userdata
---@param allocator userdata
---@return integer
function SpawnAnimation (ecs, allocator)
   local a = allocator
   local id = api.lvl.newEntity(ecs, a)
   api.lvl.addComponent(ecs, a, id, "metadata", [[{
      "archetype": "animation"
   }]])
   api.lvl.addComponent(ecs, a, id, "is_animation", nil)
   return id
end

File: src/archetypes.zig
const std = @import("std");
const Component = @import("components.zig");
const options = @import("options.zig");
const ecs = @import("ecs.zig");

pub fn createSlime(self: *ecs.ECS, a: std.mem.Allocator) !usize {
    const id = self.newEntity(a) orelse return error.EntityCapReached;
    try self.setComponent(a, id, Component.Physics{});
    try self.setComponent(a, id, Component.Hitbox{});
    try self.setComponent(a, id, Component.Health{});
    try self.setComponent(a, id, Component.WallCollisions{});
    try self.setComponent(a, id, Component.EntityCollisions{});
    try self.setComponent(a, id, Component.Metadata{ .archetype = "slime" });
    try self.setComponent(a, id, Component.Sprite{ .animation_player = .{ .animation_name = "slime" }, .styling = .{ .bob = .{} } });
    try self.setComponent(a, id, Component.Controller{});
    try self.setComponent(a, id, Component.Wanderer{});
    //try self.setComponent(a, id, Component.Light{});
    try self.setComponent(a, id, Component.MovementParticles{});
    return id;
}

pub fn createPlayer(self: *ecs.ECS, a: std.mem.Allocator) !usize {
    const id = self.newEntity(a) orelse return error.EntityCapReached;
    try self.setComponent(a, id, Component.Physics{});
    try self.setComponent(a, id, Component.EntityCollisions{});
    try self.setComponent(a, id, Component.Hitbox{});
    try self.setComponent(a, id, Component.Health{});
    try self.setComponent(a, id, Component.WallCollisions{});
    try self.setComponent(a, id, Component.Inventory{});
    try self.setComponent(a, id, Component.Metadata{ .archetype = "player" });
    try self.setComponent(a, id, Component.Sprite{ .animation_player = .{ .animation_name = "player" }, .styling = .{ .lean = .{}, .bob = .{} } });
    try self.setComponent(a, id, Component.IsPlayer{});
    try self.setComponent(a, id, Component.MovementParticles{});
    try self.setComponent(a, id, Component.Light{ .radius_in_tiles = 8 });
    return id;
}

pub fn createFireball(self: *ecs.ECS, a: std.mem.Allocator) !usize {
    const id = self.newEntity(a) orelse return error.EntityCapReached;
    try self.setComponent(a, id, Component.Physics{});
    try self.setComponent(a, id, Component.Hitbox{ .top = 0.02, .right = 0.22, .bottom = 0.22, .left = 0.02 });
    try self.setComponent(a, id, Component.WallCollisions{});
    try self.setComponent(a, id, Component.Metadata{ .archetype = "projectile" });
    try self.setComponent(a, id, Component.Sprite{ .animation_player = .{ .animation_name = "fireball" } });
    try self.setComponent(a, id, Component.Damage{});
    try self.setComponent(a, id, Component.DieWithAnimation{});
    try self.setComponent(a, id, Component.Light{ .color = .{ .x = 0.5, .y = 0.1, .z = 0.1, .a = 0.9 }, .radius_in_tiles = 2 });
    return id;
}

pub fn createParticle(self: *ecs.ECS, a: std.mem.Allocator) !usize {
    const id = self.newEntity(a) orelse return error.EntityCapReached;
    try self.setComponent(a, id, Component.Physics{});
    try self.setComponent(a, id, Component.Sprite{
        .animation_player = .{ .animation_name = "particle" },
        .z_level = .background,
        .styling = .{
            .scale = .{},
        },
    });
    try self.setComponent(a, id, Component.Metadata{ .archetype = "particle" });
    try self.setComponent(a, id, Component.Lifetime{ .milliseconds_life_remaining = 2 * 1000 });
    return id;
}
//
//pub fn createPotion(self: *ecs.ECS, a: std.mem.Allocator) !usize {
//    const id = self.newEntity(a) orelse return error.EntityCapReached;
//    try self.setComponent(a, id, Component.Physics{});
//    try self.setComponent(a, id, Component.Hitbox{});
//    try self.setComponent(a, id, Component.WallCollisions{});
//    try self.setComponent(a, id, Component.Item{
//        .animation_player = .{ .animation_name = "potion" },
//        .type_of_item = "healing_potion",
//        .action = .spawn_slime,
//    });
//    try self.setComponent(a, id, Component.Metadata{ .archetype = "item" });
//    //try self.setComponent(a, id, Component.Light{ .color = .{ .x = 0.5, .y = 0.1, .z = 0.1, .a = 0.9 }, .radius_in_tiles = 0.5 });
//    return id;
//}
